File: app/src/main/java/com/appsys/android/popularmovie/adapter/TrailerAdapter.java
package com.appsys.android.popularmovie.adapter;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.appsys.android.popularmovie.R;
import com.appsys.android.popularmovie.classes.MovieTrailer;
import com.squareup.picasso.Picasso;

import java.util.ArrayList;

/**
 * Created by shakir on 8/2/2017.
 */

public class TrailerAdapter extends RecyclerView.Adapter<TrailerAdapter.TrailerAdapterViewHolder> {

    private ArrayList<MovieTrailer> mMovies;
    private TrailerAdapterOnClickHandler mOnClickHandler;

    public TrailerAdapter(TrailerAdapterOnClickHandler onClick) {
        mOnClickHandler = onClick;
        mMovies = new ArrayList<MovieTrailer>();
    }

    public void resetMovieList() {
        mMovies.clear();
        notifyDataSetChanged();
    }

    public void setMoviesData(ArrayList<MovieTrailer> movieArray) {
        mMovies.addAll(movieArray);
        notifyDataSetChanged();
    }

    public ArrayList<MovieTrailer> getMoviesData() {
        return mMovies;
    }

    @Override
    public TrailerAdapterViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        View view = inflater.inflate(R.layout.movie_trailer_item, parent, false);
        return new TrailerAdapterViewHolder(view);
    }

    @Override
    public void onBindViewHolder(TrailerAdapterViewHolder holder, int position) {
        holder.bind(mMovies.get(position));
    }

    @Override
    public int getItemCount() {
        return mMovies.size();
    }

    public class TrailerAdapterViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        public final ImageView mImageView;
        public final TextView mTitle;

        public TrailerAdapterViewHolder(View view) {
            super(view);
            mImageView = (ImageView) view.findViewById(R.id.trailer_poster);
            mTitle = (TextView) view.findViewById(R.id.trailer_title);
            view.setOnClickListener(this);
        }

        void bind(MovieTrailer m) {
            mTitle.setText(m.getTitle());
            Picasso p = Picasso.with(mImageView.getContext());
            p.load(m.getImageUrl()).placeholder(R.mipmap.ic_launcher).error(R.mipmap.not_found).into(mImageView);
        }

        @Override
        public void onClick(View view) {
            int adapterPosition = getAdapterPosition();
            MovieTrailer m = mMovies.get(adapterPosition);
            mOnClickHandler.onClick(m);
       }
    }

    public interface TrailerAdapterOnClickHandler {
        void onClick(MovieTrailer m);
    }
}

File: app/src/main/java/com/appsys/android/popularmovie/adapter/TrailerAdapter.kt
package com.appsys.android.popularmovie.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.appsys.android.popularmovie.classes.MovieTrailer
import com.shakirfattani.course.movielisting.R
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.main.movie_trailer_item.view.*
import java.util.*

/**
 * Created by shakir on 8/2/2017.
 */
class TrailerAdapter(private val mOnClickHandler: TrailerAdapterOnClickHandler, val moviesData: ArrayList<MovieTrailer> = arrayListOf()) : RecyclerView.Adapter<TrailerAdapter.TrailerAdapterViewHolder>() {

    fun resetMovieList() {
        moviesData.clear()
        notifyDataSetChanged()
    }

    fun setMoviesData(movieArray: ArrayList<MovieTrailer>) {
        moviesData.addAll(movieArray)
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) =
            TrailerAdapterViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.movie_trailer_item, parent, false))

    override fun onBindViewHolder(holder: TrailerAdapterViewHolder, position: Int) = holder.bind(moviesData[position])

    override fun getItemCount() = moviesData.size

    interface TrailerAdapterOnClickHandler {
        fun onClick(m: MovieTrailer?)
    }

    inner class TrailerAdapterViewHolder(view: View) : RecyclerView.ViewHolder(view), View.OnClickListener {
        fun bind(m: MovieTrailer) {
            itemView.trailer_title.text = m.title
            Picasso.with(itemView.context).load(m.imageUrl).placeholder(R.mipmap.ic_launcher).error(R.mipmap.not_found).into(itemView.trailer_poster)
        }

        override fun onClick(view: View) {
            val adapterPosition = adapterPosition
            val m = moviesData[adapterPosition]
            mOnClickHandler.onClick(m)
        }

        init {
            view.setOnClickListener(this)
        }
    }
}